src/optimize/scalar/bracket.rs

```diff
@@ -0,0 +1,188 @@
+use std::cmp::Ordering::Equal;
+use std::fmt::Debug;
+use num_traits::{abs, Float, FromPrimitive, Pow, Signed};
+use thiserror::Error;
+
+#[derive(Debug, Copy, Clone)]
+pub struct Bracket<T: Float + PartialOrd + Debug> {
+	pub left: T,
+	pub f_left: T,
+	pub right: T,
+	pub f_right: T,
+	pub center: T,
+	pub f_center: T
+}
+
+#[derive(Error, Debug, Copy, Clone)]
+pub enum BracketError{
+	#[error("A bracket bound is duplicated and therefore a bracket would be invalid")]
+	DupeBoundForBracket,
+	#[error("The specified bracket does not encapsulate a dip. It might have a local minima within the bound but this is unknown")]
+	BracketNotADip
+}
+
+
+impl<T: Float + PartialOrd + Debug> Bracket<T>{
+	pub fn new(x1: T, x2: T, x3: T, func: fn(T) -> T) -> Result<Self, BracketError>{
+		let mut elements: [T; 3] = [x1, x2, x3];
+
+		elements.sort_by(|a, b| a.partial_cmp(b).unwrap_or(Equal));
+
+		Self::new_processed(elements[0], func(elements[0]), elements[1], func(elements[1]), elements[2], func(elements[2]))
+	}
+
+	pub fn new_processed(left: T, f_left: T, center: T, f_center: T, right: T, f_right: T) -> Result<Self, BracketError>{
+
+		if left == center || center == right{
+			return Err(BracketError::DupeBoundForBracket)
+		}
+
+		if !((f_center < f_left) && (f_center < f_right)) {
+			return Err(BracketError::BracketNotADip)
+		}
+
+		Ok(Bracket {
+			left,
+			f_left,
+			right,
+			f_right,
+			center,
+			f_center
+		})
+	}
+
+	pub fn longer_bound(&self, func: fn(T) -> T, ratio: T) -> Result<Bracket<T>, BracketError> {
+		let new_val = self.new_val(ratio);
+		let f_new_val = func(new_val);
+		dbg!(&new_val);
+
+		let new: Self;
+
+		if (f_new_val < self.f_center) & (new_val < self.center){
```

Comment:
Some weird indentation in this nest of else ifs. I suggest applying `rustfmt`

Reply:
Fixed in 7276fa11568a553861a4ffe912326caa638a7b4f.